py-pkg/pkgnet/function_reporter.py

```diff
@@ -0,0 +1,88 @@
+import inspect
+
+import pandas as pd
+
+from pkgnet.abstract_graph_reporter import AbstractGraphReporter
+from pkgnet.graphs import DirectedGraph
+from pkgnet.package_report import register_reporter
+from pkgnet.search_utils import (
+    get_all_modules_in_package,
+    get_fully_qualified_name,
+    get_object,
+    get_package_name,
+    safe_import_module,
+)
+
+
+@register_reporter
+class FunctionReporter(AbstractGraphReporter):
+
+    _graph_class = DirectedGraph
+
+    report_template = "tab_function_report.jinja"
+    report_slug = "function-report"
+    report_name = "Functions"
+    layout = "kamada_kawai_layout"
+
+    ### PROPERTIES ###
+
+    ### PUBLIC METHODS ###
+
+    ### PRIVATE METHODS ###
+
+    def _extract_nodes_and_edges(self):
+        if self.pkg_name is None:
+            raise AttributeError("pkg_name is not set for this reporter.")
+
+        modules = get_all_modules_in_package(self.pkg_name)
+
+        # Get all functions used in modules
+        pkg_fcns = set()
+        for module_name in modules:
+            module_obj = safe_import_module(module_name)
+            if module_obj is None:
+                continue
+            pkg_fcns |= {
+                get_fully_qualified_name(fcn_object)
+                for _, fcn_object in inspect.getmembers(module_obj, inspect.isfunction)
+                # Only want functions defined in this module, not imported functions
+                if fcn_object.__module__ == module_name
+            }
+
+        # Edges
+        dfs = []
+        for fcn_name in pkg_fcns:
+            called_fcns = self._get_called_functions(fcn_name)
+            # If function A calls function B, then A -> B
+            # A is the SOURCE and B is the TARGET
+            # This is UML dependency convention
+            dfs.append(
+                pd.DataFrame({"SOURCE": [fcn_name] * len(called_fcns), "TARGET": called_fcns})
+            )
+
+        self._edges = pd.concat(dfs, axis=0, ignore_index=True)
+
+        # Nodes
+        all_fcns = set(self.edges["SOURCE"].values) | set(self.edges["TARGET"].values)
+        self._nodes = pd.DataFrame(index=all_fcns)
+        self.nodes["package"] = self.nodes.index.map(get_package_name)
+
+    def _get_called_functions(self, fcn_name):
+        fcn_obj = get_object(fcn_name)
+        referenced_names = fcn_obj.__code__.co_names
+        module = safe_import_module(fcn_obj.__module__)
+        referenced_objs = [getattr(module, ref_name, None) for ref_name in referenced_names]
+        referenced_fcn_names = [
+            get_fully_qualified_name(obj)
+            for obj in referenced_objs
+            if obj is not None and is_package_function(obj, self.pkg_name)
+        ]
+        return referenced_fcn_names
+
+
+def is_package_function(obj, pkg_name):
```

Comment:
Are these ever used outside of `FunctionReporter`? If not I think it would make sense to make them private static methods of `FunctionReport`, like

```
@staticmethod
def _is_package_function(obj, pkg_name):
```

Reply:
So I'm a little on the fence about this one. I originally had implemented it like you suggested, but then I read some things where some people's opinions are that module-level functions are preferred over static methods (since they effectively do the same thing). 

My reasoning for making it a function is:

- It does a pretty generic thing (In this case: does this function belong to this package?), so it may be generically useful to someone without needing to care about the reporter.
- Because Python has module namespaces, you'd have to import this submodule to even get this function to show up. If you just `import pkgnet`, I don't include this function in that interface.